libsplinter/src/admin/store/diesel/operations/add_circuit.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides the "add circuit" operation for the `DieselAdminServiceStore`.
+
+use diesel::{dsl::insert_into, prelude::*, result::Error::NotFound, sql_types::Text};
+
+use super::AdminServiceStoreOperations;
+use crate::admin::store::{
+    diesel::{
+        models::{
+            CircuitMemberModel, CircuitModel, ServiceAllowedNodeModel, ServiceArgumentModel,
+            ServiceModel,
+        },
+        schema::{circuit, circuit_member, service, service_allowed_node, service_argument},
+    },
+    error::AdminServiceStoreError,
+    Circuit, CircuitNode,
+};
+
+pub(in crate::admin::store::diesel) trait AdminServiceStoreAddCircuitOperation {
+    fn add_circuit(
+        &self,
+        circuit: Circuit,
+        nodes: Vec<CircuitNode>,
+    ) -> Result<(), AdminServiceStoreError>;
+}
+
+impl<'a, C> AdminServiceStoreAddCircuitOperation for AdminServiceStoreOperations<'a, C>
+where
+    C: diesel::Connection,
+    <C as diesel::Connection>::Backend: diesel::backend::SupportsDefaultKeyword,
```

Comment:
SQLite connections don't implement `SupportsDefaultKeyword`; for other stores, we've had to implement the "add" trait separately for SQLite and Postgres.